azurelinuxagent/ga/update.py

```diff
@@ -1529,26 +1525,12 @@ def __init__(self, path=None, pkg=None, is_fast_track_goal_state=False, host=Non
             self._ensure_downloaded()
             self._ensure_loaded()
         except Exception as e:
-            if isinstance(e, ResourceGoneError):
-                raise
-
-            # The agent was improperly blacklisting versions due to a timeout
-            # encountered while downloading a later version. Errors of type
-            # socket.error are IOError, so this should provide sufficient
-            # protection against a large class of I/O operation failures.
-            if isinstance(e, IOError):
-                raise
-
-            # If we're unable to download/unpack the agent, delete the Agent directory and the zip file (if exists) to
-            # ensure we try downloading again in the next round.
+            # If we're unable to download/unpack the agent, delete the Agent directory
             try:
                 if os.path.isdir(self.get_agent_dir()):
                     shutil.rmtree(self.get_agent_dir(), ignore_errors=True)
-                if os.path.isfile(self.get_agent_pkg_path()):
-                    os.remove(self.get_agent_pkg_path())
```

Comment:
I feel like we should keep this too which gives second chance to remove incase first attempt failed or didn't get to the unzip package path.

Reply:
we still have the cleanup logic that runs periodically.

also, a new goal state would find that the ZIP is there and skip the download (and remove the ZIP after expanding it)

Reply:
I don't find the cleanup logic for agent ZIPs. When I check the vm, zips still there. I don't think we have cleanup thing. 

I'm more worry about if some reason agent gets partial zip or corrupted zip while downloading and received exception before expanding it. In that case on next goal state, we would skip the download if that zip is still in the vm.

Reply:
A bad download should be covered as well.

The previous logic was trying to delete on download or expand failures. That cleanup is done covered elsewhere: if we begin the download, start creating the file and then downloading the rest of the file fails, the WireClient.stream would delete the ZIP. If the download succeeds, but the ZIP is corrupt and expand fails, then WireClient._try_expand_zip_package would remove it.

I was mistaken in my previous comment: a new goal state does not check for the existence of the ZIP, but rather for the manifest file. If for any reason a bad ZIP is still there from a previous goal state, the new goal state would do the download and overwrite the bad ZIP.

Lastly, during agent update, UpdateHandler._purge_agents removes agent directories and ZIPs that are no longer in the goal state.

All of this maintains the same logic as before, except that some of the resposibilities have moved elsewhere as part of using the common download logic.

Reply:
Sonds like we are covering all the cases, thanks for explaining.